

<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      <link rel="icon" href="">
    
    
      
        <title>Example 7: Running PARC for clustering analysis of Covid-19 scRNA cells - PARC</title>
      
    
    
      
        
        
      
      

    
    
    
      
        
        
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
        <link rel="stylesheet" type="text/css" href="../_static/sphinx_immaterial_theme.b454d0e296e565629.min.css" />
        <link rel="stylesheet" type="text/css" href="../_static/nbsphinx-code-cells.css" />
        <link rel="stylesheet" type="text/css" href="../_static/css/custom.css" />
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="" data-md-color-primary="indigo" data-md-color-accent="">
  
    
    
      <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script>
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#Introduction" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../index.html" title="PARC" class="md-header__button md-logo" aria-label="PARC" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>
    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            PARC
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Example 7: Running PARC for clustering analysis of Covid-19 scRNA cells
            
          </span>
        </div>
      </div>
    </div>
    
      <form class="md-header__option" data-md-component="palette">
        
          
          
          <input class="md-option" data-md-color-media="" data-md-color-scheme="" data-md-color-primary="indigo" data-md-color-accent=""  aria-hidden="true"  type="radio" name="__palette" id="__palette_1">
          
        
      </form>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
          <a href="javascript:void(0)" class="md-search__icon md-icon" title="Share" aria-label="Share" data-clipboard data-clipboard-text="" data-md-component="search-share" tabindex="-1">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7 0-.24-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.15c-.05.21-.08.43-.08.66 0 1.61 1.31 2.91 2.92 2.91 1.61 0 2.92-1.3 2.92-2.91A2.92 2.92 0 0 0 18 16.08Z"/></svg>
          </a>
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/ahill187/PARC/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.2.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg>
  </div>
  <div class="md-source__repository">
    PARC
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../index.html" title="PARC" class="md-nav__button md-logo" aria-label="PARC" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>
    </a>
    PARC
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/ahill187/PARC/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.2.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg>
  </div>
  <div class="md-source__repository">
    PARC
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
  
    <li class="md-nav__item">
      <a href="../installation.html" class="md-nav__link">
        <span title="/installation.rst (reference label)" class="md-ellipsis">Installation</span>
      </a>
    </li>
  

    
      
      
      

  
  
    
  
  
    
      
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_2" type="checkbox" id="__nav_2" checked>
      
      
      
        
        <div class="md-nav__link md-nav__link--index ">
          <a href="index.html"><span title="/examples/index.rst (reference label)" class="md-ellipsis">Examples</span></a>
          
            <label for="__nav_2">
              <span class="md-nav__icon md-icon"></span>
            </label>
          
        </div>
      
      <nav class="md-nav" aria-label="Examples" data-md-level="1">
        <label class="md-nav__title" for="__nav_2">
          <span class="md-nav__icon md-icon"></span>
          <span title="/examples/index.rst (reference label)" class="md-ellipsis">Examples</span>
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
  
    <li class="md-nav__item">
      <a href="example-1.html" class="md-nav__link">
        <span title="/examples/example-1.rst (reference label)" class="md-ellipsis">Example 1:<wbr> COVID-<wbr>19 sc<wbr>RNA-<wbr>seq Data</span>
      </a>
    </li>
  

            
          
            
              
  
  
  
  
    <li class="md-nav__item">
      <a href="example-2.html" class="md-nav__link">
        <span title="/examples/example-2.rst (reference label)" class="md-ellipsis">Example 2:<wbr> Iris Dataset from sklearn</span>
      </a>
    </li>
  

            
          
            
              
  
  
  
  
    <li class="md-nav__item">
      <a href="example-3.html" class="md-nav__link">
        <span title="/examples/example-3.rst (reference label)" class="md-ellipsis">Example 3:<wbr> Digits Dataset from sklearn</span>
      </a>
    </li>
  

            
          
            
              
  
  
  
  
    <li class="md-nav__item">
      <a href="example-4.html" class="md-nav__link">
        <span title="/examples/example-4.rst (reference label)" class="md-ellipsis">Example 4:<wbr> <wbr>(mid-<wbr>scale sc<wbr>RNA-<wbr>seq):<wbr> 10X PBMC <wbr>(Zheng et al.<wbr>, 2017)</span>
      </a>
    </li>
  

            
          
            
              
  
  
  
  
    <li class="md-nav__item">
      <a href="example-5.html" class="md-nav__link">
        <span title="/examples/example-5.rst (reference label)" class="md-ellipsis">Example 5:<wbr> 10X PBMC <wbr>(Zheng et al.<wbr>, 2017) integrating scanpy pipeline</span>
      </a>
    </li>
  

            
          
            
              
  
  
  
  
    <li class="md-nav__item">
      <a href="example-6.html" class="md-nav__link">
        <span title="/examples/example-6.rst (reference label)" class="md-ellipsis">Example 6:<wbr> Large-<wbr>scale <wbr>(70K subset and 1.<wbr>1M cells) Lung Cancer cells <wbr>(multi-<wbr>ATOM imaging cytometry based features)</span>
      </a>
    </li>
  

            
          
            
              
  
  
    
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      
      <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          <span title="/examples/notebook-covid19.ipynb (reference label)" class="md-ellipsis">Example 7:<wbr> Running PARC for clustering analysis of Covid-<wbr>19 sc<wbr>RNA cells</span>
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="#" class="md-nav__link md-nav__link--active">
        <span title="/examples/notebook-covid19.ipynb (reference label)" class="md-ellipsis">Example 7:<wbr> Running PARC for clustering analysis of Covid-<wbr>19 sc<wbr>RNA cells</span>
      </a>
      
        

  

<nav class="md-nav md-nav--secondary">
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      <span title="/examples/notebook-covid19.ipynb (reference label)" class="md-ellipsis">Example 7:<wbr> Running PARC for clustering analysis of Covid-<wbr>19 sc<wbr>RNA cells</span>
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#Introduction" class="md-nav__link">
    <span title="/examples/notebook-covid19.ipynb#introduction (reference label)" class="md-ellipsis">Introduction</span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#Imports" class="md-nav__link">
    <span title="/examples/notebook-covid19.ipynb#imports (reference label)" class="md-ellipsis">Imports</span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#Download-the-Data" class="md-nav__link">
    <span title="/examples/notebook-covid19.ipynb#download-the-data (reference label)" class="md-ellipsis">Download the Data</span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#Filtering-and-pre-processing" class="md-nav__link">
    <span title="/examples/notebook-covid19.ipynb#filtering-and-pre-processing (reference label)" class="md-ellipsis">Filtering and pre-<wbr>processing</span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#Harmony-PCA-to-integrate-the-batches" class="md-nav__link">
    <span title="/examples/notebook-covid19.ipynb#harmony-pca-to-integrate-the-batches (reference label)" class="md-ellipsis">Harmony PCA to integrate the batches</span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#Run-PARC-clustering-and-get-PARC-UMAP-embedding" class="md-nav__link">
    <span title="/examples/notebook-covid19.ipynb#run-parc-clustering-and-get-parc-umap-embedding (reference label)" class="md-ellipsis">Run PARC clustering and get PARC-<wbr>UMAP embedding</span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#Marker-gene-expression-of-macrophage-and-non-macrophage-clusters" class="md-nav__link">
    <span title="/examples/notebook-covid19.ipynb#marker-gene-expression-of-macrophage-and-non-macrophage-clusters (reference label)" class="md-ellipsis">Marker gene expression of macrophage and non-<wbr>macrophage clusters</span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#Expression-of-macrophage-marker-genes" class="md-nav__link">
    <span title="/examples/notebook-covid19.ipynb#expression-of-macrophage-marker-genes (reference label)" class="md-ellipsis">Expression of macrophage marker genes</span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  
  

    
      
      
      

  
  
  
  
    <li class="md-nav__item">
      <a href="../algorithm/index.html" class="md-nav__link">
        <span title="/algorithm/index.rst (reference label)" class="md-ellipsis">Algorithm</span>
      </a>
    </li>
  

    
      
      
      

  
  
  
  
    <li class="md-nav__item">
      <a href="../parameters.html" class="md-nav__link">
        <span title="/parameters.rst (reference label)" class="md-ellipsis">Parameters and Attributes</span>
      </a>
    </li>
  

    
      
      
      

  
  
  
  
    <li class="md-nav__item">
      <a href="../dev/index.html" class="md-nav__link">
        <span title="/dev/index.rst (reference label)" class="md-ellipsis">Developers</span>
      </a>
    </li>
  

    
      
      
      

  
  
  
  
    <li class="md-nav__item">
      <a href="../references.html" class="md-nav__link">
        <span title="/references.rst (reference label)" class="md-ellipsis">References</span>
      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

  

<nav class="md-nav md-nav--secondary">
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      <span title="/examples/notebook-covid19.ipynb (reference label)" class="md-ellipsis">Example 7:<wbr> Running PARC for clustering analysis of Covid-<wbr>19 sc<wbr>RNA cells</span>
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#Introduction" class="md-nav__link">
    <span title="/examples/notebook-covid19.ipynb#introduction (reference label)" class="md-ellipsis">Introduction</span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#Imports" class="md-nav__link">
    <span title="/examples/notebook-covid19.ipynb#imports (reference label)" class="md-ellipsis">Imports</span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#Download-the-Data" class="md-nav__link">
    <span title="/examples/notebook-covid19.ipynb#download-the-data (reference label)" class="md-ellipsis">Download the Data</span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#Filtering-and-pre-processing" class="md-nav__link">
    <span title="/examples/notebook-covid19.ipynb#filtering-and-pre-processing (reference label)" class="md-ellipsis">Filtering and pre-<wbr>processing</span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#Harmony-PCA-to-integrate-the-batches" class="md-nav__link">
    <span title="/examples/notebook-covid19.ipynb#harmony-pca-to-integrate-the-batches (reference label)" class="md-ellipsis">Harmony PCA to integrate the batches</span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#Run-PARC-clustering-and-get-PARC-UMAP-embedding" class="md-nav__link">
    <span title="/examples/notebook-covid19.ipynb#run-parc-clustering-and-get-parc-umap-embedding (reference label)" class="md-ellipsis">Run PARC clustering and get PARC-<wbr>UMAP embedding</span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#Marker-gene-expression-of-macrophage-and-non-macrophage-clusters" class="md-nav__link">
    <span title="/examples/notebook-covid19.ipynb#marker-gene-expression-of-macrophage-and-non-macrophage-clusters (reference label)" class="md-ellipsis">Marker gene expression of macrophage and non-<wbr>macrophage clusters</span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#Expression-of-macrophage-marker-genes" class="md-nav__link">
    <span title="/examples/notebook-covid19.ipynb#expression-of-macrophage-marker-genes (reference label)" class="md-ellipsis">Expression of macrophage marker genes</span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset" role="main">
                  


  <a href="https://github.com/ahill187/PARC/blob/main/docs/examples/notebook-covid19.ipynb" title="Edit this page" class="md-content__button md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M10 20H6V4h7v5h5v3.1l2-2V8l-6-6H6c-1.1 0-2 .9-2 2v16c0 1.1.9 2 2 2h4v-2m10.2-7c.1 0 .3.1.4.2l1.3 1.3c.2.2.2.6 0 .8l-1 1-2.1-2.1 1-1c.1-.1.2-.2.4-.2m0 3.9L14.1 23H12v-2.1l6.1-6.1 2.1 2.1Z"/></svg>
  </a>

<h1 id="Example-7:-Running-PARC-for-clustering-analysis-of-Covid-19-scRNA-cells">Example 7: Running PARC for clustering analysis of Covid-19 scRNA cells<a class="headerlink" href="#Example-7:-Running-PARC-for-clustering-analysis-of-Covid-19-scRNA-cells" title="Permalink to this heading">¶</a></h1>
<h2 id="Introduction">Introduction<a class="headerlink" href="#Introduction" title="Permalink to this heading">¶</a></h2>
<p>Parc is a fast clustering algorithm designed to effectively cluster heterogeneity in large single cell data. We show how PARC enables downstream analysis on the recent dataset published by <a class="reference external" href="https://www.nature.com/articles/s41591-020-0901-9">Liao. et al (2020)</a>.</p>
<h2 id="Imports">Imports<a class="headerlink" href="#Imports" title="Permalink to this heading">¶</a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span><code>[3]:
</code></pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><code><span class="kn">import</span> <span class="nn">warnings</span>
<span class="n">warnings</span><span class="o">.</span><span class="n">filterwarnings</span><span class="p">(</span><span class="s1">&#39;ignore&#39;</span><span class="p">)</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">matplotlib.colors</span> <span class="k">as</span> <span class="nn">colors</span>
<span class="kn">import</span> <span class="nn">matplotlib.cm</span> <span class="k">as</span> <span class="nn">cmx</span>
<span class="kn">import</span> <span class="nn">warnings</span>
<span class="kn">import</span> <span class="nn">pathlib</span>
<span class="c1"># from numba.errors import NumbaPerformanceWarning</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">import</span> <span class="nn">parc</span>
<span class="kn">import</span> <span class="nn">harmonypy</span> <span class="k">as</span> <span class="nn">hm</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>
<span class="kn">import</span> <span class="nn">umap.umap_</span> <span class="k">as</span> <span class="nn">umap</span>
</code></pre></div>
</div>
</div>
<h2 id="Download-the-Data">Download the Data<a class="headerlink" href="#Download-the-Data" title="Permalink to this heading">¶</a></h2>
<p>The data is available on <a class="reference external" href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE145926">GEO GSE145926</a> with each of the 12 patients in a separate .h5 file. The result should be a matrix of shape (n_cells x n_genes) (108230, 33538).</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span><code>[4]:
</code></pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><code><span class="c1"># Set the directory by replacing {PATH/TO/PARC}</span>
<span class="n">DATA_DIR</span> <span class="o">=</span> <span class="s2">&quot;{PATH/TO/DATA}/&quot;</span>

<span class="n">data_files</span> <span class="o">=</span> <span class="p">[</span>
    <span class="n">pathlib</span><span class="o">.</span><span class="n">Path</span><span class="p">(</span><span class="n">DATA_DIR</span><span class="p">,</span> <span class="s2">&quot;GSM4475051_C148_filtered_feature_bc_matrix.h5&quot;</span><span class="p">),</span>
    <span class="n">pathlib</span><span class="o">.</span><span class="n">Path</span><span class="p">(</span><span class="n">DATA_DIR</span><span class="p">,</span> <span class="s2">&quot;GSM4475052_C149_filtered_feature_bc_matrix.h5&quot;</span><span class="p">),</span>
    <span class="n">pathlib</span><span class="o">.</span><span class="n">Path</span><span class="p">(</span><span class="n">DATA_DIR</span><span class="p">,</span> <span class="s2">&quot;GSM4475053_C152_filtered_feature_bc_matrix.h5&quot;</span><span class="p">),</span>
    <span class="n">pathlib</span><span class="o">.</span><span class="n">Path</span><span class="p">(</span><span class="n">DATA_DIR</span><span class="p">,</span> <span class="s2">&quot;GSM4475048_C51_filtered_feature_bc_matrix.h5&quot;</span><span class="p">),</span>
    <span class="n">pathlib</span><span class="o">.</span><span class="n">Path</span><span class="p">(</span><span class="n">DATA_DIR</span><span class="p">,</span> <span class="s2">&quot;GSM4475049_C52_filtered_feature_bc_matrix.h5&quot;</span><span class="p">),</span>
    <span class="n">pathlib</span><span class="o">.</span><span class="n">Path</span><span class="p">(</span><span class="n">DATA_DIR</span><span class="p">,</span> <span class="s2">&quot;GSM4339769_C141_filtered_feature_bc_matrix.h5&quot;</span><span class="p">),</span>
    <span class="n">pathlib</span><span class="o">.</span><span class="n">Path</span><span class="p">(</span><span class="n">DATA_DIR</span><span class="p">,</span> <span class="s2">&quot;GSM4339770_C142_filtered_feature_bc_matrix.h5&quot;</span><span class="p">),</span>
    <span class="n">pathlib</span><span class="o">.</span><span class="n">Path</span><span class="p">(</span><span class="n">DATA_DIR</span><span class="p">,</span> <span class="s2">&quot;GSM4339771_C143_filtered_feature_bc_matrix.h5&quot;</span><span class="p">),</span>
    <span class="n">pathlib</span><span class="o">.</span><span class="n">Path</span><span class="p">(</span><span class="n">DATA_DIR</span><span class="p">,</span> <span class="s2">&quot;GSM4339772_C144_filtered_feature_bc_matrix.h5&quot;</span><span class="p">),</span>
    <span class="n">pathlib</span><span class="o">.</span><span class="n">Path</span><span class="p">(</span><span class="n">DATA_DIR</span><span class="p">,</span> <span class="s2">&quot;GSM4475050_C100_filtered_feature_bc_matrix.h5&quot;</span><span class="p">),</span>
    <span class="n">pathlib</span><span class="o">.</span><span class="n">Path</span><span class="p">(</span><span class="n">DATA_DIR</span><span class="p">,</span> <span class="s2">&quot;GSM4339773_C145_filtered_feature_bc_matrix.h5&quot;</span><span class="p">),</span>
    <span class="n">pathlib</span><span class="o">.</span><span class="n">Path</span><span class="p">(</span><span class="n">DATA_DIR</span><span class="p">,</span> <span class="s2">&quot;GSM4339774_C146_filtered_feature_bc_matrix.h5&quot;</span><span class="p">)</span>
<span class="p">]</span>
</code></pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span><code>[5]:
</code></pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><br/><span></span><code><span class="n">patient_types</span> <span class="o">=</span><span class="p">[</span><span class="s1">&#39;S4&#39;</span><span class="p">,</span><span class="s1">&#39;S5&#39;</span><span class="p">,</span><span class="s1">&#39;S6&#39;</span><span class="p">,</span><span class="s1">&#39;HC1&#39;</span><span class="p">,</span><span class="s1">&#39;HC2&#39;</span><span class="p">,</span><span class="s1">&#39;M1&#39;</span><span class="p">,</span><span class="s1">&#39;M2&#39;</span><span class="p">,</span><span class="s1">&#39;S2&#39;</span><span class="p">,</span><span class="s1">&#39;M3&#39;</span><span class="p">,</span><span class="s1">&#39;HC3&#39;</span><span class="p">,</span><span class="s1">&#39;S1&#39;</span><span class="p">,</span><span class="s1">&#39;S3&#39;</span><span class="p">]</span>
<span class="n">patient_health_statuses</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;S&#39;</span><span class="p">,</span><span class="s1">&#39;S&#39;</span><span class="p">,</span><span class="s1">&#39;S&#39;</span><span class="p">,</span><span class="s1">&#39;H&#39;</span><span class="p">,</span><span class="s1">&#39;H&#39;</span><span class="p">,</span><span class="s1">&#39;M&#39;</span><span class="p">,</span><span class="s1">&#39;M&#39;</span><span class="p">,</span><span class="s1">&#39;S&#39;</span><span class="p">,</span><span class="s1">&#39;M&#39;</span><span class="p">,</span><span class="s1">&#39;H&#39;</span><span class="p">,</span><span class="s1">&#39;S&#39;</span><span class="p">,</span><span class="s1">&#39;S&#39;</span><span class="p">]</span>

<span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">patient_type</span><span class="p">,</span> <span class="n">patient_health</span><span class="p">,</span> <span class="n">file_path</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">patient_types</span><span class="p">)),</span> <span class="n">patient_types</span><span class="p">,</span> <span class="n">patient_health_statuses</span><span class="p">,</span> <span class="n">data_files</span><span class="p">):</span>
    <span class="n">ann_data_patient</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_10x_h5</span><span class="p">(</span><span class="n">file_path</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">i</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">ann_data_patient</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;patient_type&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">patient_type</span><span class="p">]</span> <span class="o">*</span> <span class="n">ann_data_patient</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">ann_data_patient</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;patient_health&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">patient_health</span><span class="p">]</span> <span class="o">*</span> <span class="n">ann_data_patient</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">ann_data_patient</span><span class="o">.</span><span class="n">var_names_make_unique</span><span class="p">()</span>
        <span class="n">ann_data</span> <span class="o">=</span> <span class="n">ann_data_patient</span>

    <span class="k">else</span><span class="p">:</span>
        <span class="n">ann_data_patient</span><span class="o">.</span><span class="n">var_names_make_unique</span><span class="p">()</span>
        <span class="n">ann_data_patient</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;patient_type&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">patient_type</span><span class="p">]</span> <span class="o">*</span> <span class="n">ann_data_patient</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">ann_data_patient</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;patient_health&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">patient_health</span><span class="p">]</span> <span class="o">*</span> <span class="n">ann_data_patient</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">ann_data</span> <span class="o">=</span> <span class="n">ann_data</span><span class="o">.</span><span class="n">concatenate</span><span class="p">(</span><span class="n">ann_data_patient</span><span class="p">,</span> <span class="n">join</span><span class="o">=</span><span class="s1">&#39;inner&#39;</span><span class="p">)</span> <span class="c1"># we want the genes in common</span>
</code></pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span><code>[6]:
</code></pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><code><span class="nb">print</span><span class="p">(</span>
    <span class="s2">&quot;Annotated data observations (first 10 rows):</span><span class="se">\n</span><span class="s2">&quot;</span>
    <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ann_data</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">10</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span>
<span class="p">)</span>
</code></pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Annotated data observations (first 10 rows):
                                         patient_type patient_health batch
AAACCTGAGAGAACAG-1-0-0-0-0-0-0-0-0-0-0-0           S4              S     0
AAACCTGAGCTGCCCA-1-0-0-0-0-0-0-0-0-0-0-0           S4              S     0
AAACCTGAGGAATCGC-1-0-0-0-0-0-0-0-0-0-0-0           S4              S     0
AAACCTGGTGGTACAG-1-0-0-0-0-0-0-0-0-0-0-0           S4              S     0
AAACCTGTCCAGAAGG-1-0-0-0-0-0-0-0-0-0-0-0           S4              S     0
AAACCTGTCCAGTAGT-1-0-0-0-0-0-0-0-0-0-0-0           S4              S     0
AAACCTGTCTGGGCCA-1-0-0-0-0-0-0-0-0-0-0-0           S4              S     0
AAACGGGAGATCCCAT-1-0-0-0-0-0-0-0-0-0-0-0           S4              S     0
AAACGGGCACGAGGTA-1-0-0-0-0-0-0-0-0-0-0-0           S4              S     0
AAACGGGGTACATCCA-1-0-0-0-0-0-0-0-0-0-0-0           S4              S     0
</pre></div></div>
</div>
<h2 id="Filtering-and-pre-processing">Filtering and pre-processing<a class="headerlink" href="#Filtering-and-pre-processing" title="Permalink to this heading">¶</a></h2>
<p>Following the filters used by Liao et.al (2020) and removing cells with mitochondrial gene proportion &gt; 0.1. After filtering the n_cells = 63753 and n_genes = 25668.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span><code>[7]:
</code></pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><code><span class="n">min_cells</span> <span class="o">=</span> <span class="mi">3</span>
<span class="n">min_genes</span> <span class="o">=</span> <span class="mi">200</span>
<span class="n">max_genes</span> <span class="o">=</span> <span class="mi">6000</span>
<span class="n">min_counts</span> <span class="o">=</span> <span class="mi">1000</span>

<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_genes</span><span class="p">(</span><span class="n">ann_data</span><span class="p">,</span> <span class="n">min_cells</span><span class="o">=</span><span class="n">min_cells</span><span class="p">)</span>  <span class="c1"># only consider genes expressed in more than min_cells</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_cells</span><span class="p">(</span><span class="n">ann_data</span><span class="p">,</span> <span class="n">min_genes</span><span class="o">=</span><span class="n">min_genes</span><span class="p">)</span> <span class="c1"># only consider cells with more than min_genes</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_cells</span><span class="p">(</span><span class="n">ann_data</span><span class="p">,</span> <span class="n">max_genes</span><span class="o">=</span><span class="n">max_genes</span><span class="p">)</span> <span class="c1"># only consider cells with less than max_cells</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_cells</span><span class="p">(</span><span class="n">ann_data</span><span class="p">,</span> <span class="n">min_counts</span><span class="o">=</span><span class="n">min_counts</span><span class="p">)</span> <span class="c1"># only consider cells with more than min_counts</span>
</code></pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span><code>[8]:
</code></pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><code><span class="n">mito_genes</span> <span class="o">=</span> <span class="n">ann_data</span><span class="o">.</span><span class="n">var_names</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;MT-&#39;</span><span class="p">)</span>
<span class="n">ann_data</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;percent_mito&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">ann_data</span><span class="p">[:,</span> <span class="n">mito_genes</span><span class="p">]</span><span class="o">.</span><span class="n">X</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">A1</span> <span class="o">/</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">ann_data</span><span class="o">.</span><span class="n">X</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">A1</span>
<span class="n">ann_data</span> <span class="o">=</span> <span class="n">ann_data</span><span class="p">[</span><span class="n">ann_data</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">percent_mito</span> <span class="o">&lt;</span> <span class="mf">0.1</span><span class="p">,</span> <span class="p">:]</span> <span class="c1"># filter cells with high mito</span>

<span class="n">ann_data</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;n_counts&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ann_data</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">A1</span> <span class="c1"># add the total counts per cell as observations-annotation to ann_data</span>

<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Shape after filtering: </span><span class="si">{</span><span class="n">ann_data</span><span class="o">.</span><span class="n">shape</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</code></pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Shape after filtering: (63753, 25668)
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span><code>[9]:
</code></pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><code><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">normalize_per_cell</span><span class="p">(</span><span class="n">ann_data</span><span class="p">,</span> <span class="n">key_n_counts</span><span class="o">=</span><span class="s1">&#39;n_counts_all&#39;</span> <span class="p">)</span>  <span class="c1"># normalize with total UMI count per cell</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">ann_data</span><span class="p">)</span>
<span class="n">ann_data</span><span class="o">.</span><span class="n">raw</span> <span class="o">=</span> <span class="n">ann_data</span>
<span class="n">adata_beforeHVG</span> <span class="o">=</span> <span class="n">ann_data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
</code></pre></div>
</div>
</div>
<h2 id="Harmony-PCA-to-integrate-the-batches">Harmony PCA to integrate the batches<a class="headerlink" href="#Harmony-PCA-to-integrate-the-batches" title="Permalink to this heading">¶</a></h2>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span><code>[10]:
</code></pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><code><span class="n">n_comps_pca</span> <span class="o">=</span> <span class="mi">50</span>
<span class="n">n_top_genes</span> <span class="o">=</span> <span class="mi">2000</span>

<span class="c1"># select HVG</span>
<span class="n">filter_result</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_genes_dispersion</span><span class="p">(</span>
    <span class="n">ann_data</span><span class="o">.</span><span class="n">X</span><span class="p">,</span>
    <span class="n">flavor</span><span class="o">=</span><span class="s1">&#39;cell_ranger&#39;</span><span class="p">,</span>
    <span class="n">n_top_genes</span><span class="o">=</span><span class="n">n_top_genes</span><span class="p">,</span>
    <span class="n">log</span><span class="o">=</span><span class="kc">False</span>
<span class="p">)</span> <span class="c1"># select highly-variable genes</span>

<span class="nb">print</span><span class="p">(</span><span class="n">filter_result</span><span class="p">,</span> <span class="n">filter_result</span><span class="o">.</span><span class="n">gene_subset</span><span class="p">)</span>

<span class="n">ann_data</span> <span class="o">=</span> <span class="n">ann_data</span><span class="p">[:,</span> <span class="n">filter_result</span><span class="o">.</span><span class="n">gene_subset</span><span class="p">]</span>  <span class="c1"># subset the genes</span>

<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">scale</span><span class="p">(</span><span class="n">ann_data</span><span class="p">,</span> <span class="n">max_value</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>  <span class="c1"># scale to unit variance and shift to zero mean. Clip values exceeding standard deviation 10.</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">ann_data</span><span class="p">,</span> <span class="n">svd_solver</span><span class="o">=</span><span class="s1">&#39;arpack&#39;</span><span class="p">,</span> <span class="n">n_comps</span><span class="o">=</span><span class="n">n_comps_pca</span><span class="p">)</span>
</code></pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[(False, 7.0834276e-04, 0.6848484 ,  0.27154154)
 (False, 7.4105184e-05, 0.6147516 ,  0.4668431 )
 (False, 2.5824214e-05, 0.9647063 ,  1.7608302 ) ...
 ( True, 1.3939296e-03, 3.886037  , 32.046425  )
 ( True, 2.6764904e-04, 1.3681211 ,  4.5427337 )
 (False, 7.9755653e-03, 0.61826074,  1.0558165 )] [False False False ...  True  True False]
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span><code>[11]:
</code></pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><code><span class="n">df_meta</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>
<span class="n">df_meta</span><span class="p">[</span><span class="s1">&#39;patient_type&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ann_data</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;patient_type&#39;</span><span class="p">]</span>
<span class="n">harmony_out</span> <span class="o">=</span> <span class="n">hm</span><span class="o">.</span><span class="n">run_harmony</span><span class="p">(</span><span class="n">ann_data</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_pca&#39;</span><span class="p">],</span> <span class="n">df_meta</span><span class="p">,</span> <span class="s1">&#39;patient_type&#39;</span><span class="p">)</span>
<span class="n">res</span> <span class="o">=</span> <span class="n">harmony_out</span><span class="o">.</span><span class="n">Z_corr</span><span class="o">.</span><span class="n">T</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Size of harmony corrected output: </span><span class="si">{</span><span class="n">res</span><span class="o">.</span><span class="n">shape</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</code></pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
2024-09-02 16:18:35,124 - harmonypy - INFO - Computing initial centroids with sklearn.KMeans...
2024-09-02 16:18:52,087 - harmonypy - INFO - sklearn.KMeans initialization complete.
2024-09-02 16:18:52,350 - harmonypy - INFO - Iteration 1 of 10
2024-09-02 16:19:16,735 - harmonypy - INFO - Iteration 2 of 10
2024-09-02 16:19:37,594 - harmonypy - INFO - Iteration 3 of 10
2024-09-02 16:19:58,838 - harmonypy - INFO - Converged after 3 iterations
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Size of harmony corrected output: (63753, 50)
</pre></div></div>
</div>
<h2 id="Run-PARC-clustering-and-get-PARC-UMAP-embedding">Run PARC clustering and get PARC-UMAP embedding<a class="headerlink" href="#Run-PARC-clustering-and-get-PARC-UMAP-embedding" title="Permalink to this heading">¶</a></h2>
<p>We construct the UMAP embedding my providing UMAP that KNN graph constructed in PARC. The PARC-UMAP implementation is runtime-efficient, and has significantly lower RAM usage.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span><code>[12]:
</code></pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><code><span class="n">parc_model</span> <span class="o">=</span> <span class="n">parc</span><span class="o">.</span><span class="n">PARC</span><span class="p">(</span><span class="n">res</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">42</span><span class="p">)</span>
<span class="n">parc_model</span><span class="o">.</span><span class="n">run_parc</span><span class="p">()</span>
<span class="n">ann_data</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;parc&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">cluster</span><span class="p">)</span> <span class="k">for</span> <span class="n">cluster</span> <span class="ow">in</span> <span class="n">parc_model</span><span class="o">.</span><span class="n">y_data_pred</span><span class="p">]</span>

<span class="n">marker_genes</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;macrophages&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;CD68&#39;</span><span class="p">,</span><span class="s1">&#39;FCN1&#39;</span><span class="p">,</span><span class="s1">&#39;SPP1&#39;</span><span class="p">,</span><span class="s1">&#39;FABP4&#39;</span><span class="p">],</span>
    <span class="s2">&quot;neutrophils&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;FCGR3B&#39;</span><span class="p">],</span>
    <span class="s2">&quot;mDC&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;CD1C&#39;</span><span class="p">,</span> <span class="s1">&#39;CLEC9A&#39;</span><span class="p">],</span>
    <span class="s2">&quot;pDC&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;LILRA4&#39;</span><span class="p">],</span>
    <span class="s2">&quot;NK&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;KLRD1&#39;</span><span class="p">],</span> <span class="s2">&quot;T-cell&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;CD3D&#39;</span><span class="p">],</span>
    <span class="s2">&quot;B-cell&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;CD19&#39;</span><span class="p">,</span><span class="s1">&#39;MS4A1&#39;</span><span class="p">,</span><span class="s1">&#39;IGHD&#39;</span><span class="p">,</span> <span class="s1">&#39;CD22&#39;</span><span class="p">],</span>
    <span class="s2">&quot;plasma&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;IGHG4&#39;</span><span class="p">],</span>  <span class="c1"># CD19 doesnt show up for B</span>
    <span class="s2">&quot;epithel&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;TPPP3&#39;</span><span class="p">,</span> <span class="s1">&#39;KRT18&#39;</span><span class="p">]</span>
<span class="p">}</span>

<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Plot cluster average expression of marker genes&#39;</span><span class="p">)</span>
<span class="n">ax_mat</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">matrixplot</span><span class="p">(</span><span class="n">ann_data</span><span class="p">,</span> <span class="n">marker_genes</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="s1">&#39;parc&#39;</span><span class="p">)</span>
<span class="n">graph</span> <span class="o">=</span> <span class="n">parc_model</span><span class="o">.</span><span class="n">create_knn_graph</span><span class="p">()</span>
<span class="c1"># umap_embedding= parc_model.run_umap_hnsw(</span>
<span class="c1">#     x_data=res,</span>
<span class="c1">#     graph=graph,</span>
<span class="c1">#     random_state_seed=1</span>
<span class="c1"># )</span>
<span class="n">umap_embedding</span> <span class="o">=</span> <span class="n">umap</span><span class="o">.</span><span class="n">UMAP</span><span class="p">(</span><span class="n">n_components</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">res</span><span class="p">)</span>
</code></pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[MESSAGE]: Input data has shape 63753 (samples) x 50 (features)
[MESSAGE]: Creating knn_struct...
[MESSAGE]: Starting local pruning based on Euclidean distance metric at 3 standard deviations above the mean
[MESSAGE]: Starting global pruning...
[MESSAGE]: Creating graph with 924364 edges and 63753 nodes...
[MESSAGE]: Starting Leiden community detection...
[MESSAGE]: Leiden algorithm find partition: partition type = ModularityVertexPartition
[MESSAGE]:
Community 0 is not too large and has size:
13077 &lt;= large_community_factor * n_samples = 0.4 * 63753 = 25501.2
Skipping large community expansion.
[MESSAGE]: Starting small community detection...
[MESSAGE]: Community 27 is a small community with population 2
[MESSAGE]: Community 28 is a small community with population 2
[MESSAGE]: Community 29 is a small community with population 2
[MESSAGE]: Community 30 is a small community with population 2
[MESSAGE]: Community 31 is a small community with population 2
[MESSAGE]: Community 32 is a small community with population 1
[MESSAGE]: Community 33 is a small community with population 1
[MESSAGE]: Community 34 is a small community with population 1
[MESSAGE]: Community 35 is a small community with population 1
[MESSAGE]: Community 36 is a small community with population 1
[MESSAGE]: Community 37 is a small community with population 1
[MESSAGE]: Community 38 is a small community with population 1
[MESSAGE]: Community 39 is a small community with population 1
[MESSAGE]: Community 40 is a small community with population 1
[MESSAGE]: Community 41 is a small community with population 1
[MESSAGE]: Community 42 is a small community with population 1
[MESSAGE]: Community 43 is a small community with population 1
[MESSAGE]: Community 44 is a small community with population 1
[MESSAGE]: Community 45 is a small community with population 1
[MESSAGE]: Community 46 is a small community with population 1
[MESSAGE]: Community 47 is a small community with population 1
[MESSAGE]: Community 48 is a small community with population 1
[MESSAGE]: Community 49 is a small community with population 1
[MESSAGE]: Community 50 is a small community with population 1
[MESSAGE]: Community 51 is a small community with population 1
[MESSAGE]: Community 52 is a small community with population 1
[MESSAGE]: Community 53 is a small community with population 1
[MESSAGE]: Community 54 is a small community with population 1
[MESSAGE]: Community 55 is a small community with population 1
[MESSAGE]: Community 56 is a small community with population 1
[MESSAGE]: Community 57 is a small community with population 1
[MESSAGE]: Community 58 is a small community with population 1
[MESSAGE]: Community 59 is a small community with population 1
[MESSAGE]: Community 60 is a small community with population 1
[MESSAGE]: Community 61 is a small community with population 1
[MESSAGE]: Community 62 is a small community with population 1
[MESSAGE]: Community 63 is a small community with population 1
[MESSAGE]: Community 64 is a small community with population 1
[MESSAGE]: Community 65 is a small community with population 1
[MESSAGE]: Community 66 is a small community with population 1
[MESSAGE]: Community 67 is a small community with population 1
[MESSAGE]: Community 68 is a small community with population 1
[MESSAGE]: Community 69 is a small community with population 1
[MESSAGE]: Community 70 is a small community with population 1
[MESSAGE]: Community 71 is a small community with population 1
[MESSAGE]: Community 72 is a small community with population 1
[MESSAGE]: Community 73 is a small community with population 1
[MESSAGE]: Community 74 is a small community with population 1
[MESSAGE]: Community 75 is a small community with population 1
[MESSAGE]: Community 76 is a small community with population 1
[MESSAGE]: Community 77 is a small community with population 1
[MESSAGE]: Community 78 is a small community with population 1
[MESSAGE]: Community 79 is a small community with population 1
[MESSAGE]: Community 80 is a small community with population 1
[MESSAGE]: Community 81 is a small community with population 1
[MESSAGE]: Community 82 is a small community with population 1
[MESSAGE]: Community 83 is a small community with population 1
[MESSAGE]: Community 84 is a small community with population 1
[MESSAGE]: Community 85 is a small community with population 1
[MESSAGE]: Community 86 is a small community with population 1
[MESSAGE]: Community 87 is a small community with population 1
[MESSAGE]: Community 88 is a small community with population 1
[MESSAGE]: Community 89 is a small community with population 1
[MESSAGE]: Community 90 is a small community with population 1
[MESSAGE]: Community 91 is a small community with population 1
[MESSAGE]: Community 92 is a small community with population 1
[MESSAGE]: Community 93 is a small community with population 1
[MESSAGE]: Community 94 is a small community with population 1
[MESSAGE]: Community 95 is a small community with population 1
[MESSAGE]: Community 96 is a small community with population 1
[MESSAGE]: Community 97 is a small community with population 1
[MESSAGE]: Community 98 is a small community with population 1
[MESSAGE]: Community 99 is a small community with population 1
[MESSAGE]: Community 100 is a small community with population 1
[MESSAGE]: Community 101 is a small community with population 1
[MESSAGE]: Community 102 is a small community with population 1
[MESSAGE]: Community 103 is a small community with population 1
[MESSAGE]: Community 104 is a small community with population 1
[MESSAGE]: Community 105 is a small community with population 1
[MESSAGE]: Community 106 is a small community with population 1
[MESSAGE]: Community 107 is a small community with population 1
[MESSAGE]: Community 108 is a small community with population 1
[MESSAGE]: Community 109 is a small community with population 1
[MESSAGE]: Community 110 is a small community with population 1
[MESSAGE]: Community 111 is a small community with population 1
[MESSAGE]: Community 112 is a small community with population 1
[MESSAGE]: Community 113 is a small community with population 1
[MESSAGE]: Community 114 is a small community with population 1
[MESSAGE]: Community 115 is a small community with population 1
[MESSAGE]: Community 116 is a small community with population 1
[MESSAGE]: Community 117 is a small community with population 1
[MESSAGE]: Community 118 is a small community with population 1
[MESSAGE]: Community 119 is a small community with population 1
[MESSAGE]: Community 120 is a small community with population 1
[MESSAGE]: Community 121 is a small community with population 1
[MESSAGE]: Community 122 is a small community with population 1
[MESSAGE]: Community 123 is a small community with population 1
[MESSAGE]: Community 124 is a small community with population 1
[MESSAGE]: Community 125 is a small community with population 1
[MESSAGE]: Community 126 is a small community with population 1
[MESSAGE]: Community 127 is a small community with population 1
[MESSAGE]: Community 128 is a small community with population 1
[MESSAGE]: Community 129 is a small community with population 1
[MESSAGE]: Community 130 is a small community with population 1
[MESSAGE]: Community 131 is a small community with population 1
[MESSAGE]: Community 132 is a small community with population 1
[MESSAGE]: Community 133 is a small community with population 1
[MESSAGE]: Community 134 is a small community with population 1
[MESSAGE]: Community 135 is a small community with population 1
[MESSAGE]: Community 136 is a small community with population 1
[MESSAGE]: Community 137 is a small community with population 1
[MESSAGE]: Community 138 is a small community with population 1
[MESSAGE]: Community 139 is a small community with population 1
[MESSAGE]: Community 140 is a small community with population 1
[MESSAGE]: Community 141 is a small community with population 1
[MESSAGE]: Community 142 is a small community with population 1
[MESSAGE]: Community 143 is a small community with population 1
[MESSAGE]: Community 144 is a small community with population 1
[MESSAGE]: Community 145 is a small community with population 1
[MESSAGE]: Community 146 is a small community with population 1
[MESSAGE]: Community 147 is a small community with population 1
[MESSAGE]: Community 148 is a small community with population 1
[MESSAGE]: Community 149 is a small community with population 1
[MESSAGE]: Community 150 is a small community with population 1
[MESSAGE]: Community 151 is a small community with population 1
[MESSAGE]: Community 152 is a small community with population 1
[MESSAGE]: Community 153 is a small community with population 1
[MESSAGE]: Community 154 is a small community with population 1
[MESSAGE]: Community 155 is a small community with population 1
[MESSAGE]: Community 156 is a small community with population 1
[MESSAGE]: Community 157 is a small community with population 1
[MESSAGE]: Community 158 is a small community with population 1
[MESSAGE]: Community 159 is a small community with population 1
[MESSAGE]: Community 160 is a small community with population 1
[MESSAGE]: Community 161 is a small community with population 1
[MESSAGE]: Community 162 is a small community with population 1
[MESSAGE]: Community 163 is a small community with population 1
[MESSAGE]: Community 164 is a small community with population 1
[MESSAGE]: Community 165 is a small community with population 1
[MESSAGE]: Community 166 is a small community with population 1
[MESSAGE]: Community 167 is a small community with population 1
[MESSAGE]: Community 168 is a small community with population 1
[MESSAGE]: Community 169 is a small community with population 1
[MESSAGE]: Community 170 is a small community with population 1
[MESSAGE]: Community 171 is a small community with population 1
[MESSAGE]: Community 172 is a small community with population 1
[MESSAGE]: Community 173 is a small community with population 1
[MESSAGE]: Community 174 is a small community with population 1
[MESSAGE]: Community 175 is a small community with population 1
[MESSAGE]: Community 176 is a small community with population 1
[MESSAGE]: Community 177 is a small community with population 1
[MESSAGE]: Community 178 is a small community with population 1
[MESSAGE]: Community 179 is a small community with population 1
[MESSAGE]: Community 180 is a small community with population 1
[MESSAGE]: Community 181 is a small community with population 1
[MESSAGE]: Community 182 is a small community with population 1
[MESSAGE]: Community 183 is a small community with population 1
[MESSAGE]: Community 184 is a small community with population 1
[MESSAGE]: Community 185 is a small community with population 1
[MESSAGE]: Community 186 is a small community with population 1
[MESSAGE]: Community 187 is a small community with population 1
[MESSAGE]: Community 188 is a small community with population 1
[MESSAGE]: Community 189 is a small community with population 1
[MESSAGE]: Community 190 is a small community with population 1
[MESSAGE]: Community 191 is a small community with population 1
[MESSAGE]: Community 192 is a small community with population 1
[MESSAGE]: Community 193 is a small community with population 1
[MESSAGE]: Community 194 is a small community with population 1
[MESSAGE]: Community 195 is a small community with population 1
[MESSAGE]: Community 196 is a small community with population 1
[MESSAGE]: Community 197 is a small community with population 1
[MESSAGE]: Community 198 is a small community with population 1
[MESSAGE]: Community 199 is a small community with population 1
[MESSAGE]: Community 200 is a small community with population 1
[MESSAGE]: Community 201 is a small community with population 1
[MESSAGE]: Community 202 is a small community with population 1
[MESSAGE]: Community 203 is a small community with population 1
[MESSAGE]: Community 204 is a small community with population 1
[MESSAGE]: Community 205 is a small community with population 1
[MESSAGE]: Community 206 is a small community with population 1
[MESSAGE]: Community 207 is a small community with population 1
[MESSAGE]: Community 208 is a small community with population 1
[MESSAGE]: Community 209 is a small community with population 1
[MESSAGE]: Community 210 is a small community with population 1
[MESSAGE]: Community 211 is a small community with population 1
[MESSAGE]: Community 212 is a small community with population 1
[MESSAGE]: Community 213 is a small community with population 1
[MESSAGE]: Community 214 is a small community with population 1
[MESSAGE]: Community 215 is a small community with population 1
[MESSAGE]: Community 216 is a small community with population 1
[MESSAGE]: Community 217 is a small community with population 1
[MESSAGE]: Community 218 is a small community with population 1
[MESSAGE]: Community 219 is a small community with population 1
[MESSAGE]: Community 220 is a small community with population 1
[MESSAGE]: Community 221 is a small community with population 1
[MESSAGE]: Community 222 is a small community with population 1
[MESSAGE]: Community 223 is a small community with population 1
[MESSAGE]: Community 224 is a small community with population 1
[MESSAGE]: Community 225 is a small community with population 1
[MESSAGE]: Community 226 is a small community with population 1
[MESSAGE]: Community 227 is a small community with population 1
[MESSAGE]: Community 228 is a small community with population 1
[MESSAGE]: Community 229 is a small community with population 1
[MESSAGE]: Community 230 is a small community with population 1
[MESSAGE]: Community 231 is a small community with population 1
[MESSAGE]: Community 232 is a small community with population 1
[MESSAGE]: Community 233 is a small community with population 1
[MESSAGE]: Community 234 is a small community with population 1
[MESSAGE]: Community 235 is a small community with population 1
[MESSAGE]: Community 236 is a small community with population 1
[MESSAGE]: Community 237 is a small community with population 1
[MESSAGE]: Community 238 is a small community with population 1
[MESSAGE]: Community 239 is a small community with population 1
[MESSAGE]: Community 240 is a small community with population 1
[MESSAGE]: Community 241 is a small community with population 1
[MESSAGE]: Community 242 is a small community with population 1
[MESSAGE]: Community 243 is a small community with population 1
[MESSAGE]: Community 244 is a small community with population 1
[MESSAGE]: Community 245 is a small community with population 1
[MESSAGE]: Community 246 is a small community with population 1
[MESSAGE]: Community 247 is a small community with population 1
[MESSAGE]: Community 248 is a small community with population 1
[MESSAGE]: Community 249 is a small community with population 1
[MESSAGE]: Community 250 is a small community with population 1
[MESSAGE]: Community 251 is a small community with population 1
[MESSAGE]: Community labels and sizes: [(0, 13169), (1, 9751), (2, 5639), (3, 5584), (4, 5087), (5, 3218), (6, 3192), (7, 2859), (8, 2689), (9, 2539), (10, 1211), (11, 1127), (12, 1061), (13, 1060), (14, 957), (15, 926), (16, 884), (17, 876), (18, 679), (19, 333), (20, 244), (21, 187), (22, 153), (23, 142), (24, 105), (25, 45), (26, 36)]
[MESSAGE]: Time elapsed to run PARC: 19.9 seconds
Plot cluster average expression of marker genes
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/examples_notebook-covid19_15_1.png" src="../_images/examples_notebook-covid19_15_1.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[MESSAGE]: a: 1.5769434604035877, b: 0.8950608780665811, spread: 1.0, dist: 0.1
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
0it [00:00, ?it/s]
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span><code>[13]:
</code></pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><code><span class="n">marker_genes</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;macrophages&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;FCN1&#39;</span><span class="p">,</span><span class="s1">&#39;SPP1&#39;</span><span class="p">,</span><span class="s1">&#39;FABP4&#39;</span><span class="p">,</span><span class="s1">&#39;CD68&#39;</span><span class="p">],</span>
    <span class="s2">&quot;neutrophils&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;FCGR3B&#39;</span><span class="p">],</span>
    <span class="s2">&quot;mDC&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;CD1C&#39;</span><span class="p">,</span> <span class="s1">&#39;CLEC9A&#39;</span><span class="p">],</span>
    <span class="s2">&quot;pDC&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;LILRA4&#39;</span><span class="p">],</span>
    <span class="s2">&quot;NK&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;KLRD1&#39;</span><span class="p">],</span>
    <span class="s2">&quot;T-cell&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;CD3D&#39;</span><span class="p">],</span>
    <span class="s2">&quot;B-cell&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;MS4A1&#39;</span><span class="p">,</span><span class="s1">&#39;IGHD&#39;</span><span class="p">,</span> <span class="s1">&#39;CD22&#39;</span><span class="p">],</span>
    <span class="s2">&quot;plasma&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;IGHG4&#39;</span><span class="p">],</span>  <span class="c1"># CD19 doesnt show up for B</span>
    <span class="s2">&quot;epithel&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;TPPP3&#39;</span><span class="p">,</span> <span class="s1">&#39;KRT18&#39;</span><span class="p">],</span>
    <span class="c1"># &quot;macrophages&quot;: [&#39;FCN1&#39;,&#39;SPP1&#39;,&#39;FABP4&#39;,&#39;CD68&#39;]</span>
<span class="p">}</span>

<span class="k">for</span> <span class="n">var_name</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;ENSG00000129226&#39;</span><span class="p">,</span><span class="s1">&#39;Cd68&#39;</span><span class="p">,</span> <span class="s1">&#39;CD68&#39;</span><span class="p">,</span> <span class="s1">&#39;SCARD1&#39;</span><span class="p">,</span><span class="s1">&#39;GP110&#39;</span><span class="p">,</span><span class="s1">&#39;LAMP4&#39;</span><span class="p">,</span><span class="s1">&#39;Gp110&#39;</span><span class="p">,</span><span class="s1">&#39;SPP1&#39;</span><span class="p">]:</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">var_name</span><span class="si">}</span><span class="s2"> in var_names: </span><span class="si">{</span><span class="n">var_name</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">ann_data</span><span class="o">.</span><span class="n">var_names</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Plot cluster average expression of marker genes&#39;</span><span class="p">)</span>
<span class="n">ax_mat</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">matrixplot</span><span class="p">(</span><span class="n">ann_data</span><span class="p">,</span> <span class="n">marker_genes</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="s1">&#39;parc&#39;</span><span class="p">)</span>
</code><br/></pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
ENSG00000129226 in var_names: False
Cd68 in var_names: False
CD68 in var_names: False
SCARD1 in var_names: False
GP110 in var_names: False
LAMP4 in var_names: False
Gp110 in var_names: False
SPP1 in var_names: True
Plot cluster average expression of marker genes
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/examples_notebook-covid19_16_1.png" src="../_images/examples_notebook-covid19_16_1.png" />
</div>
</div>
<h2 id="Marker-gene-expression-of-macrophage-and-non-macrophage-clusters">Marker gene expression of macrophage and non-macrophage clusters<a class="headerlink" href="#Marker-gene-expression-of-macrophage-and-non-macrophage-clusters" title="Permalink to this heading">¶</a></h2>
<p>Compute average cluster level gene expressions and sort those with high cd68 as macrophage clusters Identify the cluster composition in terms of H, M and S to label each cluster as predominantly H, M or S</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span><code>[20]:
</code></pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><code><span class="n">clustermap_marker_</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;Group 1&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;S100A8&#39;</span><span class="p">,</span> <span class="s1">&#39;FCN1&#39;</span><span class="p">,</span><span class="s1">&#39;CD14&#39;</span><span class="p">],</span>
    <span class="s1">&#39;Group 2&#39;</span><span class="p">:[</span><span class="s1">&#39;CCL2&#39;</span><span class="p">,</span> <span class="s1">&#39;CCL3&#39;</span><span class="p">,</span> <span class="s1">&#39;CXCL10&#39;</span><span class="p">],</span>
    <span class="s1">&#39;Group 1-2&#39;</span><span class="p">:[</span><span class="s1">&#39;STAT1&#39;</span><span class="p">,</span><span class="s1">&#39;STAT2&#39;</span><span class="p">],</span>
    <span class="s1">&#39;Group 3&#39;</span><span class="p">:[</span><span class="s1">&#39;SPP1&#39;</span><span class="p">,</span><span class="s1">&#39;A2M&#39;</span><span class="p">,</span> <span class="s1">&#39;GPR183&#39;</span><span class="p">,</span><span class="s1">&#39;CCL13&#39;</span><span class="p">,</span><span class="s1">&#39;CREB1&#39;</span><span class="p">,</span><span class="s1">&#39;TFEB&#39;</span><span class="p">,</span><span class="s1">&#39;NR1H3&#39;</span><span class="p">,</span> <span class="s1">&#39;PPARA&#39;</span><span class="p">],</span>
    <span class="s1">&#39;Group 4&#39;</span><span class="p">:[</span><span class="s1">&#39;FABP4&#39;</span><span class="p">,</span><span class="s1">&#39;APOC1&#39;</span><span class="p">,</span> <span class="s1">&#39;MARCO&#39;</span><span class="p">,</span><span class="s1">&#39;PPARG&#39;</span><span class="p">,</span> <span class="s1">&#39;CEBPB&#39;</span><span class="p">]</span>
<span class="p">}</span>
<span class="n">clustermap_marker_genes</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;Healthy&#39;</span><span class="p">:[</span><span class="s1">&#39;FABP4&#39;</span><span class="p">],</span>
    <span class="s1">&#39;Moderate&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;FABP4&#39;</span><span class="p">],</span>
    <span class="s1">&#39;Severe&#39;</span><span class="p">:[</span><span class="s1">&#39;FCN1&#39;</span><span class="p">,</span><span class="s1">&#39;SPP1&#39;</span><span class="p">,</span><span class="s1">&#39;TFEB&#39;</span><span class="p">,</span><span class="s1">&#39;PPARG&#39;</span><span class="p">,</span> <span class="s1">&#39;CEBPB&#39;</span><span class="p">,</span><span class="s1">&#39;CREB1&#39;</span><span class="p">,</span><span class="s1">&#39;NR1H3&#39;</span><span class="p">,</span> <span class="s1">&#39;PPARA&#39;</span><span class="p">]</span>
<span class="p">}</span>

<span class="c1"># marker genes for non-macrophage clusters</span>
<span class="n">marker_genes</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;mDC&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;CD1C&#39;</span><span class="p">,</span> <span class="s1">&#39;CLEC9A&#39;</span><span class="p">],</span>
    <span class="s2">&quot;pDC&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;LILRA4&#39;</span><span class="p">],</span>
    <span class="s2">&quot;NK&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;KLRD1&#39;</span><span class="p">],</span>
    <span class="s1">&#39;T-cell&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;CD3D&#39;</span><span class="p">],</span>
    <span class="s1">&#39;B-cell&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;MS4A1&#39;</span><span class="p">,</span><span class="s1">&#39;IGHD&#39;</span><span class="p">,</span> <span class="s1">&#39;CD22&#39;</span><span class="p">],</span>
    <span class="s1">&#39;plasma&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;IGHG4&#39;</span><span class="p">],</span>  <span class="c1"># CD19 doesnt show up for B</span>
    <span class="s1">&#39;epithel&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;TPPP3&#39;</span><span class="p">,</span> <span class="s1">&#39;KRT18&#39;</span><span class="p">]</span>
<span class="p">}</span>

<span class="n">ann_data</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;macrophage&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ann_data</span><span class="o">.</span><span class="n">raw</span><span class="p">[:,</span> <span class="s1">&#39;CD68&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span>

<span class="n">df_adata</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">ann_data</span><span class="o">.</span><span class="n">X</span><span class="p">,</span> <span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="n">ann_data</span><span class="o">.</span><span class="n">var_names</span><span class="p">)])</span>
<span class="n">df_adata</span><span class="p">[</span><span class="s1">&#39;parc&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">ann_data</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;parc&#39;</span><span class="p">]]</span>
<span class="n">df_adata</span><span class="p">[</span><span class="s1">&#39;cd68&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">ann_data</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;macrophage&#39;</span><span class="p">]]</span> <span class="c1">#CD68 is not in the filtered HVG genes, but in adata.RAW</span>

<span class="n">df_adata</span> <span class="o">=</span> <span class="n">df_adata</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;parc&#39;</span><span class="p">,</span> <span class="n">as_index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>
<span class="n">df_adata</span><span class="p">[</span><span class="s1">&#39;macrophage&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df_adata</span><span class="p">[</span><span class="s1">&#39;cd68&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">df_adata</span><span class="p">[</span><span class="s1">&#39;cd68&#39;</span><span class="p">]</span><span class="o">+</span><span class="mf">0.2</span><span class="o">*</span><span class="n">np</span><span class="o">.</span><span class="n">std</span><span class="p">(</span><span class="n">df_adata</span><span class="p">[</span><span class="s1">&#39;cd68&#39;</span><span class="p">])))</span>

<span class="n">macrophage_cluster_list</span> <span class="o">=</span> <span class="n">df_adata</span><span class="p">[</span><span class="n">df_adata</span><span class="p">[</span><span class="s1">&#39;macrophage&#39;</span><span class="p">]</span><span class="o">==</span><span class="kc">True</span><span class="p">][</span><span class="s1">&#39;parc&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>

<span class="n">patient_health_list</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">ann_data</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;patient_health&#39;</span><span class="p">]]</span>
<span class="n">parc_health</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">empty</span><span class="p">([</span><span class="nb">len</span><span class="p">(</span><span class="n">parc_model</span><span class="o">.</span><span class="n">y_data_pred</span><span class="p">),</span> <span class="mi">1</span><span class="p">],</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">object</span><span class="p">)</span>
<span class="n">health_dict</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;H&quot;</span><span class="p">:[],</span> <span class="s2">&quot;S&quot;</span><span class="p">:[],</span> <span class="s2">&quot;M&quot;</span><span class="p">:[]}</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">set</span><span class="p">(</span><span class="n">parc_model</span><span class="o">.</span><span class="n">y_data_pred</span><span class="p">):</span>
    <span class="n">loc_i</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="n">parc_model</span><span class="o">.</span><span class="n">y_data_pred</span><span class="p">)</span><span class="o">==</span><span class="n">i</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">ll</span><span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="n">patient_health_list</span><span class="p">)[</span><span class="n">loc_i</span><span class="p">])</span>
    <span class="n">mode_i</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">ll</span><span class="p">),</span> <span class="n">key</span><span class="o">=</span><span class="n">ll</span><span class="o">.</span><span class="n">count</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">mode_i</span> <span class="o">==</span> <span class="s1">&#39;S&#39;</span><span class="p">:</span>
        <span class="n">parc_health</span><span class="p">[</span><span class="n">loc_i</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;S&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
        <span class="n">health_dict</span><span class="p">[</span><span class="s2">&quot;S&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">mode_i</span> <span class="o">==</span> <span class="s1">&#39;M&#39;</span><span class="p">:</span>
        <span class="n">parc_health</span><span class="p">[</span><span class="n">loc_i</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;M&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
        <span class="n">health_dict</span><span class="p">[</span><span class="s2">&quot;M&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">mode_i</span> <span class="o">==</span> <span class="s1">&#39;H&#39;</span><span class="p">:</span>
        <span class="n">parc_health</span><span class="p">[</span><span class="n">loc_i</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;H&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
        <span class="n">health_dict</span><span class="p">[</span><span class="s2">&quot;H&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>

<span class="n">parc_health</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">parc_health</span><span class="o">.</span><span class="n">flatten</span><span class="p">())</span>

<span class="n">ann_data</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;health_parc&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">parc_health</span><span class="p">]</span>
<span class="n">macrophage_cluster_bool</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">parc_model</span><span class="o">.</span><span class="n">y_data_pred</span><span class="p">:</span>
    <span class="k">if</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="ow">in</span> <span class="n">macrophage_cluster_list</span><span class="p">:</span>
        <span class="n">macrophage_cluster_bool</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">macrophage_cluster_bool</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>

<span class="n">ann_data</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;macro_clus&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">macrophage_cluster_bool</span>
<span class="n">adata_macro</span> <span class="o">=</span> <span class="n">ann_data</span><span class="p">[</span><span class="n">ann_data</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;macro_clus&#39;</span><span class="p">]</span><span class="o">==</span><span class="kc">True</span><span class="p">]</span>
<span class="n">adata_not_macro</span> <span class="o">=</span> <span class="n">ann_data</span><span class="p">[</span><span class="n">ann_data</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;macro_clus&#39;</span><span class="p">]</span><span class="o">!=</span><span class="kc">True</span><span class="p">]</span>

<span class="n">new_health_parc</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">adata_macro</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;health_parc&#39;</span><span class="p">]]</span>

<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">matrixplot</span><span class="p">(</span>
    <span class="n">adata_macro</span><span class="p">,</span>
    <span class="n">clustermap_marker_genes</span><span class="p">,</span>
    <span class="n">groupby</span><span class="o">=</span><span class="s1">&#39;health_parc&#39;</span><span class="p">,</span>
    <span class="n">vmax</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
    <span class="n">vmin</span><span class="o">=-</span><span class="mi">1</span><span class="p">,</span>
    <span class="n">dendrogram</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">matrixplot</span><span class="p">(</span>
    <span class="n">adata_not_macro</span><span class="p">,</span>
    <span class="n">marker_genes</span><span class="p">,</span>
    <span class="n">groupby</span><span class="o">=</span><span class="s1">&#39;health_parc&#39;</span><span class="p">,</span>
    <span class="n">vmax</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
    <span class="n">vmin</span><span class="o">=-</span><span class="mi">1</span><span class="p">,</span>
    <span class="n">dendrogram</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>
</code></pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
WARNING: dendrogram data not found (using key=dendrogram_health_parc). Running `sc.tl.dendrogram` with default parameters. For fine tuning it is recommended to run `sc.tl.dendrogram` independently.
WARNING: Groups are not reordered because the `groupby` categories and the `var_group_labels` are different.
categories: H1, H2, H4, etc.
var_group_labels: Healthy, Moderate, Severe
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/examples_notebook-covid19_18_1.png" src="../_images/examples_notebook-covid19_18_1.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
WARNING: dendrogram data not found (using key=dendrogram_health_parc). Running `sc.tl.dendrogram` with default parameters. For fine tuning it is recommended to run `sc.tl.dendrogram` independently.
WARNING: Groups are not reordered because the `groupby` categories and the `var_group_labels` are different.
categories: H15, M6, S7, etc.
var_group_labels: mDC, pDC, NK, etc.
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/examples_notebook-covid19_18_3.png" src="../_images/examples_notebook-covid19_18_3.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span><code>[21]:
</code></pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><code><span class="n">color_views</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;FCN1&#39;</span><span class="p">,</span><span class="s1">&#39;SPP1&#39;</span><span class="p">,</span><span class="s1">&#39;FABP4&#39;</span><span class="p">]</span>

<span class="n">fig</span><span class="p">,</span> <span class="n">axs</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">24</span><span class="p">,</span> <span class="mi">12</span><span class="p">))</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Patient Health&quot;</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;PARC&quot;</span><span class="p">)</span>

<span class="n">color_dict</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;H&quot;</span><span class="p">:</span><span class="s1">&#39;Green&#39;</span><span class="p">,</span><span class="s2">&quot;S&quot;</span><span class="p">:</span><span class="s1">&#39;Red&#39;</span><span class="p">,</span><span class="s2">&quot;M&quot;</span><span class="p">:</span><span class="s1">&#39;Yellow&#39;</span><span class="p">}</span>

<span class="n">cmaps</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;S&#39;</span><span class="p">:</span> <span class="n">colors</span><span class="o">.</span><span class="n">LinearSegmentedColormap</span><span class="o">.</span><span class="n">from_list</span><span class="p">(</span>
        <span class="s1">&#39;name&#39;</span><span class="p">,</span> <span class="n">plt</span><span class="o">.</span><span class="n">get_cmap</span><span class="p">(</span><span class="s1">&#39;Reds&#39;</span><span class="p">)(</span><span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="mf">0.2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">100</span><span class="p">))</span>
    <span class="p">),</span>
    <span class="s1">&#39;H&#39;</span><span class="p">:</span> <span class="n">colors</span><span class="o">.</span><span class="n">LinearSegmentedColormap</span><span class="o">.</span><span class="n">from_list</span><span class="p">(</span>
        <span class="s1">&#39;name&#39;</span><span class="p">,</span> <span class="n">plt</span><span class="o">.</span><span class="n">get_cmap</span><span class="p">(</span><span class="s1">&#39;Greens&#39;</span><span class="p">)(</span><span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="mf">0.2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">100</span><span class="p">))</span>
    <span class="p">),</span>
    <span class="s1">&#39;M&#39;</span><span class="p">:</span> <span class="n">colors</span><span class="o">.</span><span class="n">LinearSegmentedColormap</span><span class="o">.</span><span class="n">from_list</span><span class="p">(</span>
        <span class="s1">&#39;name&#39;</span><span class="p">,</span> <span class="n">plt</span><span class="o">.</span><span class="n">get_cmap</span><span class="p">(</span><span class="s2">&quot;Wistia&quot;</span><span class="p">)(</span><span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="mf">0.1</span><span class="p">,</span> <span class="mf">0.4</span><span class="p">,</span> <span class="mi">100</span><span class="p">))</span>
    <span class="p">)</span>
<span class="p">}</span>

<span class="n">dict_color</span> <span class="o">=</span> <span class="p">{}</span>
<span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">uniq</span> <span class="ow">in</span> <span class="n">health_dict</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
    <span class="n">cnorm</span> <span class="o">=</span> <span class="n">colors</span><span class="o">.</span><span class="n">Normalize</span><span class="p">(</span><span class="n">vmin</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">uniq</span><span class="p">))</span>
    <span class="n">scalar_map</span> <span class="o">=</span> <span class="n">cmx</span><span class="o">.</span><span class="n">ScalarMappable</span><span class="p">(</span><span class="n">norm</span><span class="o">=</span><span class="n">cnorm</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="n">cmaps</span><span class="p">[</span><span class="n">key</span><span class="p">])</span>

    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">uniq</span><span class="p">:</span>
        <span class="n">indx</span> <span class="o">=</span> <span class="n">ann_data</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;parc&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
        <span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span>
            <span class="n">umap_embedding</span><span class="p">[</span><span class="n">indx</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span>
            <span class="n">umap_embedding</span><span class="p">[</span><span class="n">indx</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span>
            <span class="n">color</span><span class="o">=</span><span class="n">scalar_map</span><span class="o">.</span><span class="n">to_rgba</span><span class="p">(</span><span class="n">i</span><span class="p">),</span>
            <span class="n">label</span><span class="o">=</span><span class="n">key</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">),</span>
            <span class="n">alpha</span><span class="o">=</span><span class="mf">0.4</span><span class="p">,</span>
            <span class="n">s</span><span class="o">=</span><span class="mi">4</span>
        <span class="p">)</span>

        <span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">umap_embedding</span><span class="p">[</span><span class="n">indx</span><span class="p">,</span><span class="mi">0</span><span class="p">]),</span><span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">umap_embedding</span><span class="p">[</span><span class="n">indx</span><span class="p">,</span><span class="mi">1</span><span class="p">]),</span> <span class="n">key</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">))</span>
        <span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>

    <span class="n">indx</span> <span class="o">=</span> <span class="n">ann_data</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;patient_health&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">key</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span>
        <span class="n">umap_embedding</span><span class="p">[</span><span class="n">indx</span><span class="p">,</span><span class="mi">0</span><span class="p">],</span>
        <span class="n">umap_embedding</span><span class="p">[</span><span class="n">indx</span><span class="p">,</span><span class="mi">1</span><span class="p">],</span>
        <span class="n">color</span><span class="o">=</span><span class="n">color_dict</span><span class="p">[</span><span class="n">key</span><span class="p">],</span>
        <span class="n">label</span><span class="o">=</span><span class="n">key</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">),</span>
        <span class="n">alpha</span><span class="o">=</span><span class="mf">0.4</span><span class="p">,</span>
        <span class="n">s</span><span class="o">=</span><span class="mi">4</span>
    <span class="p">)</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</code><br/></pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/examples_notebook-covid19_19_0.png" src="../_images/examples_notebook-covid19_19_0.png" />
</div>
</div>
<h2 id="Expression-of-macrophage-marker-genes">Expression of macrophage marker genes<a class="headerlink" href="#Expression-of-macrophage-marker-genes" title="Permalink to this heading">¶</a></h2>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span><code>[23]:
</code></pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><code><span class="n">color_views</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;FCN1&#39;</span><span class="p">,</span> <span class="s1">&#39;SPP1&#39;</span><span class="p">,</span> <span class="s1">&#39;FABP4&#39;</span><span class="p">]</span>

<span class="n">fig</span><span class="p">,</span> <span class="n">axs</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span> <span class="p">,</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">24</span><span class="p">,</span><span class="mi">24</span><span class="p">))</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">3</span><span class="p">):</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span>
        <span class="n">umap_embedding</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">],</span>
        <span class="n">umap_embedding</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">],</span>
        <span class="n">c</span><span class="o">=</span><span class="n">ann_data</span><span class="p">[:,</span> <span class="n">color_views</span><span class="p">[</span><span class="n">i</span><span class="p">]]</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">flatten</span><span class="p">(),</span>
        <span class="n">alpha</span><span class="o">=</span><span class="mf">0.4</span><span class="p">,</span>
        <span class="n">s</span><span class="o">=</span><span class="mi">3</span>
    <span class="p">)</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="n">color_views</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>

<span class="n">color_views_categ</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;patient_type&#39;</span><span class="p">,</span> <span class="s1">&#39;patient_health&#39;</span><span class="p">,</span> <span class="s1">&#39;parc&#39;</span><span class="p">]</span>
<span class="n">color_dict</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;H&quot;</span><span class="p">:</span><span class="s1">&#39;Green&#39;</span><span class="p">,</span><span class="s2">&quot;S&quot;</span><span class="p">:</span><span class="s1">&#39;Red&#39;</span><span class="p">,</span><span class="s2">&quot;M&quot;</span><span class="p">:</span><span class="s1">&#39;Yellow&#39;</span><span class="p">}</span>

<span class="k">for</span> <span class="n">i_ax</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">2</span><span class="p">):</span>
    <span class="n">uniq</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">ann_data</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">color_views_categ</span><span class="p">[</span><span class="n">i_ax</span><span class="p">]]))</span>
    <span class="c1"># Set the color map to match the number of species</span>
    <span class="n">cnorm</span> <span class="o">=</span> <span class="n">colors</span><span class="o">.</span><span class="n">Normalize</span><span class="p">(</span><span class="n">vmin</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">uniq</span><span class="p">))</span>
    <span class="n">scalar_map</span> <span class="o">=</span> <span class="n">cmx</span><span class="o">.</span><span class="n">ScalarMappable</span><span class="p">(</span><span class="n">norm</span><span class="o">=</span><span class="n">cnorm</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;RdYlGn&#39;</span><span class="p">)</span>
    <span class="c1"># Plot each species</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">uniq</span><span class="p">)):</span>
        <span class="n">indx</span> <span class="o">=</span> <span class="n">ann_data</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">color_views_categ</span><span class="p">[</span><span class="n">i_ax</span><span class="p">]]</span> <span class="o">==</span> <span class="n">uniq</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
        <span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="n">i_ax</span><span class="p">]</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span>
            <span class="n">umap_embedding</span><span class="p">[</span><span class="n">indx</span><span class="p">,</span><span class="mi">0</span><span class="p">],</span>
            <span class="n">umap_embedding</span><span class="p">[</span><span class="n">indx</span><span class="p">,</span><span class="mi">1</span><span class="p">],</span>
            <span class="n">color</span><span class="o">=</span><span class="n">scalar_map</span><span class="o">.</span><span class="n">to_rgba</span><span class="p">(</span><span class="n">i</span><span class="p">),</span>
            <span class="n">label</span><span class="o">=</span><span class="n">uniq</span><span class="p">[</span><span class="n">i</span><span class="p">],</span>
            <span class="n">alpha</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span>
            <span class="n">s</span><span class="o">=</span><span class="mi">3</span>
        <span class="p">)</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="n">i_ax</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="n">color_views_categ</span><span class="p">[</span><span class="n">i_ax</span><span class="p">])</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="n">i_ax</span><span class="p">]</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>


<span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">uniq</span> <span class="ow">in</span> <span class="n">health_dict</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
    <span class="n">cnorm</span> <span class="o">=</span> <span class="n">colors</span><span class="o">.</span><span class="n">Normalize</span><span class="p">(</span><span class="n">vmin</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">uniq</span><span class="p">))</span>
    <span class="n">scalar_map</span> <span class="o">=</span> <span class="n">cmx</span><span class="o">.</span><span class="n">ScalarMappable</span><span class="p">(</span><span class="n">norm</span><span class="o">=</span><span class="n">cnorm</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="n">cmaps</span><span class="p">[</span><span class="n">key</span><span class="p">])</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">uniq</span><span class="p">:</span>
        <span class="n">indx</span> <span class="o">=</span> <span class="n">ann_data</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;parc&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
        <span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span>
            <span class="n">umap_embedding</span><span class="p">[</span><span class="n">indx</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span>
            <span class="n">umap_embedding</span><span class="p">[</span><span class="n">indx</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span>
            <span class="n">color</span><span class="o">=</span><span class="n">scalar_map</span><span class="o">.</span><span class="n">to_rgba</span><span class="p">(</span><span class="n">i</span><span class="p">),</span>
            <span class="n">label</span><span class="o">=</span><span class="n">key</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">),</span>
            <span class="n">alpha</span><span class="o">=</span><span class="mf">0.4</span><span class="p">,</span>
            <span class="n">s</span><span class="o">=</span><span class="mi">3</span>
        <span class="p">)</span>
        <span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="n">color_views_categ</span><span class="p">[</span><span class="n">i_ax</span><span class="p">])</span>
        <span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">umap_embedding</span><span class="p">[</span><span class="n">indx</span><span class="p">,</span> <span class="mi">0</span><span class="p">]),</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">umap_embedding</span><span class="p">[</span><span class="n">indx</span><span class="p">,</span> <span class="mi">1</span><span class="p">]),</span><span class="n">key</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">))</span>
        <span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
    <span class="n">indx</span> <span class="o">=</span> <span class="n">ann_data</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;patient_health&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">key</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span>
        <span class="n">umap_embedding</span><span class="p">[</span><span class="n">indx</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span>
        <span class="n">umap_embedding</span><span class="p">[</span><span class="n">indx</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span>
        <span class="n">color</span><span class="o">=</span><span class="n">color_dict</span><span class="p">[</span><span class="n">key</span><span class="p">],</span>
        <span class="n">label</span><span class="o">=</span><span class="n">key</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">),</span>
        <span class="n">alpha</span><span class="o">=</span><span class="mf">0.4</span><span class="p">,</span>
        <span class="n">s</span><span class="o">=</span><span class="mi">3</span>
    <span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</code><br/></pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/examples_notebook-covid19_21_0.png" src="../_images/examples_notebook-covid19_21_0.png" />
</div>
</div>






                
              </article>
            </div>
          
          
  <script>var tabs=__md_get("__tabs");if(Array.isArray(tabs))e:for(var set of document.querySelectorAll(".tabbed-set")){var tab,labels=set.querySelector(".tabbed-labels");for(tab of tabs)for(var label of labels.getElementsByTagName("label"))if(label.innerText.trim()===tab){var input=document.getElementById(label.htmlFor);input.checked=!0;continue e}}</script>

        </div>
        
          <a href="#" class="md-top md-icon" data-md-component="top" hidden>
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg>
            Back to top
          </a>
        
      </main>
      
        <footer class="md-footer">
  
    
    <nav class="md-footer__inner md-grid" aria-label="Footer" >
      
        
        <a href="example-6.html" class="md-footer__link md-footer__link--prev" aria-label="Previous: Example 6: Large-scale (70K subset and 1.1M cells) Lung Cancer cells (multi-ATOM imaging cytometry based features)" rel="prev">
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
          </div>
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Previous
              </span>
              Example 6: Large-scale (70K subset and 1.1M cells) Lung Cancer cells (multi-ATOM imaging cytometry based features)
            </div>
          </div>
        </a>
      
      
        
        <a href="../algorithm/index.html" class="md-footer__link md-footer__link--next" aria-label="Next: Algorithm" rel="next">
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Next
              </span>
              Algorithm
            </div>
          </div>
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg>
          </div>
        </a>
      
    </nav>
  
  
  
  <div class="md-footer-meta md-typeset">
    
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-footer-copyright__highlight">
        &#169; Copyright 2022, Shobana Stassen.
        
    </div>
  
    Created using
    <a href="https://www.sphinx-doc.org/" target="_blank" rel="noopener">Sphinx</a>
    7.0.0.
     and
    <a href="https://github.com/jbms/sphinx-immaterial/" target="_blank" rel="noopener">Sphinx-Immaterial</a>
  
</div>
      
        <div class="md-social">
  
    
    
    <a href="https://github.com/ahill187/PARC/" target="_blank" rel="noopener" title="Source on github.com" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.2.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg>
    </a>
  
</div>
      
    </div>
    
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    <script id="__config" type="application/json">{"base": "..", "features": ["navigation.expand", "navigation.sections", "navigation.top", "search.share", "toc.follow", "toc.sticky", "content.tabs.link", "announce.dismiss", "content.code.annotate"], "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version.title": "Select version"}}</script>
    
      
        <script src="../_static/sphinx_immaterial_theme.f9d9eeeb247ace16c.min.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
        <script id="MathJax-script" src="../_static/mathjax/tex-mml-chtml.js"></script>
    
  </body>
</html>